datacube/scripts/product.py

```diff
@@ -104,30 +112,106 @@ def update_dataset_types(index, allow_unsafe, allow_exclusive_lock, dry_run, fil
                                                                                  len(safe_changes)))
     sys.exit(failures)
 
-
-@product.command('list')
-@ui.pass_datacube()
-def list_products(dc):
+def build_product_list(index):
+    lstdct = []
+    for line in index.products.search():
+        info = dataset_type_to_row(line)
+        lstdct.append(info)
+    return lstdct
+
+def _write_csv(index):
+    writer = csv.DictWriter(sys.stdout, ['id', 'name', 'description',\
+                            'ancillary_quality', 'latgqa_cep90', 'product_type',\
+                            'gqa_abs_iterative_mean_xy', 'gqa_ref_source', 'sat_path',\
+                            'gqa_iterative_stddev_xy', 'time', 'sat_row', 'orbit', 'gqa',\
+                            'instrument', 'gqa_abs_xy', 'crs', 'resolution', 'tile_size',\
+                            'spatial_dimensions'], extrasaction='ignore')
+    writer.writeheader()
+
+    def add_first_name(row):
+        names_ = row['name']
+        row['name'] = names_ if names_ else None
+        return row
+
+    writer.writerows(add_first_name(row) for row in index)
+
+def _write_yaml(index):
     """
-    List products that are defined in the generic index.
+    Dump yaml data with support for OrderedDicts.
+
+    Allows for better human-readability of output: such as dataset ID field first, sources last.
+
+    (Ordered dicts are output identically to normal yaml dicts: their order is purely for readability)
     """
-    products = dc.list_products()
+
+    # We can't control how many ancestors this dumper API uses.
+    # pylint: disable=too-many-ancestors
+    class OrderedDumper(yaml.SafeDumper):
```

Comment:
This class should be moved out from inside these methods, and shared between the different command line scripts. It needs a more descriptive name, and should include all the features that are required in different places, ie. Safe, and able to represent `OrderedDicts`,  Dates and Ranges

Either create a `common.py` file inside the `scripts/` directory. Or put it inside `datacube/utils/`.

It's safe for the single